feat: add e2e tests for trezor ERC721 NFT deployment and management.

dawnseeker8 opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**
Link to this plan [Trezor ERC712](https://github.com/MetaMask/accounts-planning/issues/968)
This PR has implemented a e2e tests for trezor ERC721 NFT deployment and management.
<!--
Write a short description of the changes included in this pull request, also include relevant motivation and context. Have in mind the following questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/35383?quickstart=1)

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and accurately)
-->

CHANGELOG entry:

## **Related issues**

Fixes: https://github.com/MetaMask/accounts-planning/issues/968

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

dawnseeker8 committed:
feat: consolidate token polling and batch RPC requests

This commit consolidates polling for native and ERC20 tokens into a single RPC request, streamlining the process and reducing the number of RPC calls. Additionally, all RPC requests across account addresses are now batched through the `multicall3` contract, enhancing performance.

dawnseeker8 committed:
refactor: update test suite description for Trezor hardware

This commit changes the test suite description from 'Ledger Hardware' to 'Trezor Hardware' in the Trezor ERC-721 specification test file, ensuring clarity and accuracy in the test context.

dawnseeker8 committed:
Merge branch 'main' into feat/trezor-erc721-e2e

dawnseeker8 committed:
Merge branch 'main' into feat/trezor-erc721-e2e

dawnseeker8 committed:
Merge branch 'main' into feat/trezor-erc721-e2e

dawnseeker8 committed:
Merge branch 'main' into feat/trezor-erc721-e2e

dawnseeker8 committed:
Merge branch 'main' into feat/trezor-erc721-e2e

dawnseeker8 committed:
Merge branch 'main' into feat/trezor-erc721-e2e

dawnseeker8 committed:
Merge branch 'main' into feat/trezor-erc721-e2e

dawnseeker8 committed:
Merge branch 'main' into feat/trezor-erc721-e2e

dawnseeker8 committed:
Merge branch 'main' into feat/trezor-erc721-e2e

dawnseeker8 committed:
Merge branch 'main' into feat/trezor-erc721-e2e

dawnseeker8 committed:
Move trezor erc 721 to trezor folder.

dawnseeker8 committed:
Merge branch 'main' into feat/trezor-erc721-e2e

seaona reviewed (commented):
Could we split this into 4 tests:
- deploy ERC
- mint
- approve
- setApprovalForAll


This way, each test is validating a specific functionality and they are isolated from one another (one can fail while the rest pass, etC)

for the first one, leave it as is, the rest of specs can use the fixture to inject the contract: see this example https://github.com/MetaMask/metamask-extension/blob/main/test/e2e/tests/confirmations/transactions/transaction-decoding-redesign.spec.ts#L21

seaona commented:
Spec is failing on FF. It seems the dialog was not opened, though I can see how the tx was added in the wallet. This should be debugged further:

> Failure on testcase: 'Trezor Hardware can perform actions on an ERC-721 token', for more information see the artifacts tab in CI
> NoSuchWindowError: Unable to locate window: 7a8b038c-1111-41f3-89c1-d3e1f04faa15


<img width="1152" height="785" alt="test-failure-screenshot-1" src="https://github.com/user-attachments/assets/098d7d2c-10bd-4bce-8c66-32190b279c99" />


<img width="1152" height="785" alt="test-failure-screenshot-2" src="https://github.com/user-attachments/assets/c799c2f3-5e0f-4c10-a504-ed6cac965a01" />

dawnseeker8 committed:
Merge branch 'main' into feat/trezor-erc721-e2e

dawnseeker8 committed:
Merge branch 'main' into feat/trezor-erc721-e2e

dawnseeker8 committed:
Merge branch 'main' into feat/trezor-erc721-e2e

dawnseeker8 committed:
test: Enhance Trezor ERC-721 e2e tests with deployment, minting, and approval actions

- Added tests for deploying, minting, and approving ERC-721 tokens using Trezor hardware wallets.
- Refactored existing test structure to improve clarity and maintainability.
- Integrated smart contract and DApp URL into test fixtures for better configuration.
- Ensured proper validation of account balances and transaction confirmations throughout the test cases.

dawnseeker8 committed:
test: Refactor Trezor ERC-721 e2e tests for contract deployment and NFT minting

- Introduced helper functions for deploying ERC-721 contracts and minting NFTs to enhance test readability and maintainability.
- Updated test cases to utilize the new functions, streamlining the process of deploying and interacting with ERC-721 tokens.
- Ensured proper validation of transaction activities and NFT display in the user interface.

dawnseeker8 committed:
Rewrite NFT approve function tests due to error to run SmartContract.NFTs

seaona reviewed (approved):
LGTM
